### PR TITLE
Tentative solution to pytree

### PR DESCRIPTION
**Pull request by mavenlin**
We may move the discussions on pytree here, there are some future PRs that would depend on a systematic solution of this.

### CODE REVIEW

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -198,6 +199,30 @@ def _coerce(x):
     )
 
 
+def _tree_coerce(x):
+  # The while loop here may look scary.
+  # Because there are no reliable ways to determine whether a type has been registered with jax, it could cause repeated registrations.
+  # If we're comfortable with jax._src._registry, an alternative is to tree_map in torch first and obtain all the custom nodes from the spec.
+  # Then we register them with jax if they are not already registered.
```
[the remark this answers is not in the document]
Comment (mavenlin) [reply torch2jax/__init__.py:206]: There is no overhead when all nodes have been registered. But when there are nested new nodes to be registered, the while loop could run longer.

**Comment by samuela** [comment]
ah this is a cute idea. i like it if it works! so iiuc the idea is that we replace https://github.com/samuela/torch2jax/blob/bee09743c2df90e24c9ccf08150392fe74f12576/torch2jax/__init__.py#L1174-L1175 with `_tree_coerce` and then it automatically handles pytorch pytree types as we encounter them? i wonder if this will work inside of `jax.jit`?

File: torch2jax/__init__.py [hunk 2]
```diff
@@ -11,9 +13,14 @@
...
 from torch.utils._pytree import tree_map as torch_tree_map
+from torch.utils._pytree import tree_structure as torch_tree_structure
 
 # so that __getitem__ & __setitem__ with mixed keys of int / tensor could work
-torch_register_pytree_node(slice, lambda s: ((s.start, s.stop, s.step), None), lambda values, ctx: slice(*values))
+torch_register_pytree_node(
+  slice,
+  lambda s: ((s.start, s.stop, s.step), None),
+  lambda values, ctx: slice(*values),
+)
```
Comment (samuela) [inline torch2jax/__init__.py:23]: iiuc this is just formatting? if so, let's keep the git blame clean and leave this as is

File: torch2jax/__init__.py [hunk 3]
```diff
@@ -211,6 +218,35 @@ def _coerce(x):
...
+    specs.extend(spec.children_specs)
+    if spec.is_leaf():
+      continue
+    if spec.type in jax._src.tree_util._registry:
+      continue  # already registered
+    node = torch.utils._pytree.SUPPORTED_NODES.get(spec.type, None)
+    if node is None:
+      continue
+    if is_dataclass(spec.type):
+      jax.tree_util.register_dataclass(spec.type)
```
Comment (samuela) [inline torch2jax/__init__.py:236]: is there an advantage to doing this instead of `jax.tree_util.register_pytree_node`?

File: torch2jax/__init__.py [hunk 4]
```diff
@@ -211,6 +218,35 @@ def _coerce(x):
...
+    else:
+      jax.tree_util.register_pytree_node(
+        spec.type,
+        node.flatten_fn,
+        lambda a, b, n=node: n.unflatten_fn(b, a),  # torch & jax have different order of args
+        # flatten_with_keys_func=node.flatten_with_keys_fn,
+        # TODO: flatten_with_keys has some problem as jax & torch has different class for key
+        # without registering it seems to just work as well.
```
Comment (samuela) [inline torch2jax/__init__.py:244]:
```suggestion
```
flatten_with_keys_func is undocumented in jax and not referenced in any of the examples. i say we just skip it entirely

File: torch2jax/__init__.py [hunk 5]
```diff
@@ -211,6 +218,35 @@ def _coerce(x):
...
+    node = torch.utils._pytree.SUPPORTED_NODES.get(spec.type, None)
+    if node is None:
+      continue
+    if is_dataclass(spec.type):
+      jax.tree_util.register_dataclass(spec.type)
+    else:
+      jax.tree_util.register_pytree_node(
+        spec.type,
+        node.flatten_fn,
+        lambda a, b, n=node: n.unflatten_fn(b, a),  # torch & jax have different order of args
```
Comment (samuela) [inline torch2jax/__init__.py:241]:
```suggestion
        lambda a, b: node.unflatten_fn(b, a),  # torch & jax have different order of args
```
docs suggest that this function takes exactly two arguments

File: tests/test_core.py [hunk 6]
```diff
@@ -55,6 +56,30 @@ def test_ones_like():
...
+
+  def f():
+    t1 = torch.return_types.max([1, torch.ones(3)])
+    t2 = torch.return_types.topk([3.0, t1])
+    t3 = A(torch.ones(3, 2))
+    t4 = B(t3, 7)
+    return (t2, t3, t4)
+
+  tests = [forward_test]
+  t2j_function_test(f, [], tests=tests)
```
Comment (samuela) [inline tests/test_core.py:80]: I'm worried that this test is missing a case: `forward_test` always calls the non-`jit` version of `f` before the `jit` version of `f`. that means we're not currently testing the case where we are calling `jit(f)` first. I'm concerned that `register_pytree_node` may not work in the jit context, and that we're just getting lucky bc we're calling the non-jit `f` first which does the registration for us. 

let's add a test that only calls the jit version to ensure that it works that way as well

**Review by samuela — COMMENTED** [review]
this contains a lot of formatting changes that have no impact on functionality. let's remove those before merging

File: torch2jax/__init__.py (hunk 5 above)
Comment (mavenlin) [reply torch2jax/__init__.py:241]: The trick here is to capture the node at the time of closure creation, otherwise the value of the node will change.

File: torch2jax/__init__.py (hunk 4 above)
Comment (mavenlin) [reply torch2jax/__init__.py:244]: Sounds good

File: torch2jax/__init__.py (hunk 3 above)
Comment (mavenlin) [reply torch2jax/__init__.py:236]: I actually don't know, without specializing it also works for the test cases.

File: torch2jax/__init__.py (hunk 2 above)
Comment (mavenlin) [reply torch2jax/__init__.py:23]: I thought it was ruff, could be that I accidentally added a comma.

File: tests/test_core.py (hunk 6 above)
Comment (mavenlin) [reply tests/test_core.py:87]: Test added, confirms it works with jit. Jit will always run the python code first for tracing, and the registration part is pure python, which jax will execute as it is.

File: torch2jax/__init__.py (hunk 5 above)
Comment (samuela) [reply torch2jax/__init__.py]: i'm not sure i follow... the lambda will capture `node` in its closure by value not by name. what would be a test case that would make this fail?

Comment (samuela) [reply torch2jax/__init__.py]: wow, that is... terrifying! TIL
